File: shared/src/main/java/uk/co/unclealex/hammers/calendar/shared/model/Competition.java
/**
 * Copyright 2010 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 * @author unclealex72
 *
 */
package uk.co.unclealex.hammers.calendar.shared.model;

import java.util.Comparator;
import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;

public enum Competition {
	PREM("Premiership", true, "PREM"),
	LGCP("League Cup", false, "LGCP"),
	FACP("FA Cup", false, "FACP"),
	FLC("Championship", true, "FLC", "FLD1"),
	FLCPO("Play-Offs", false, "FLD1 P/O");
	
	private static SortedMap<String, Competition> COMPETITIONS_BY_TOKEN;
	static {
		Comparator<String> longestFirstComparator = new Comparator<String>() {
			@Override
			public int compare(String s1, String s2) {
				int cmp = s2.length() - s1.length();
				return cmp == 0?s1.compareTo(s2):cmp;
			}
		};
		COMPETITIONS_BY_TOKEN = new TreeMap<String, Competition>(longestFirstComparator);
		for (Competition competition : Competition.values()) {
			for (String token : competition.getTokens()) {
				COMPETITIONS_BY_TOKEN.put(token, competition);
			}
		}
	}
	
	public static Competition findByToken(String token) throws IllegalArgumentException {
		for (Map.Entry<String, Competition> entry : COMPETITIONS_BY_TOKEN.entrySet()) {
			if (token.startsWith(entry.getKey())) {
				return entry.getValue();
			}
		}
		throw new IllegalArgumentException(token + " is not a valid competition token.");
	}
	
	private String[] i_tokens;
	private boolean i_league;
	private String i_name;
	
	private Competition(String name, boolean league, String... tokens) {
		i_name = name;
		i_league = league;
		i_tokens = tokens;
	}

	public String getName() {
		return i_name;
	}
	
	public String[] getTokens() {
		return i_tokens;
	}

	/**
	 * @return True if this is a league competition, false otherwise.
	 */
	public boolean isLeague() {
		return i_league;
	}
}

File: shared/src/main/scala/models/Competition.scala
/**
 * Copyright 2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation  extends Competition(0, ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0  extends Competition(0, the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package models

import enumeratum._
import json.JsonEnum

import scala.collection.immutable

/**
 * The different competitions that West Ham have taken part in.
 *
 * @author alex
 *
 */
sealed trait Competition extends EnumEntry {
  /**
   * The English name of this competition.
   */
  val name: String
  /**
   * The tokens that are used within the West Ham website to identify this competition.
   */
  val tokens: Seq[String]

  /**
   * A flag that determines whether this competition is a primary English league competition or not.
   */
  def isLeague: Boolean

}

sealed abstract class AbstractCompetition(val name: String, val isLeague: Boolean, val tokens: String*) extends Competition

/**
 * League based competitions.
 */
sealed abstract class LeagueCompetition(override val name: String, override val tokens: String*) extends AbstractCompetition(name, true, tokens: _*)

/**
 * Cup based competitions.
 */
sealed abstract class CupCompetition(override val name: String, override val tokens: String*) extends AbstractCompetition(name, false, tokens: _*)

object Competition extends JsonEnum[Competition] {

  val values: immutable.IndexedSeq[Competition] = findValues

  /**
   * The FA Premiership.
   */
  case object PREM extends LeagueCompetition("Premiership", "Barclays Premier League", "Premier League", "premier-league")

  /**
   * The League Cup.
   */
  case object LGCP extends CupCompetition(
    "League Cup", "English Capital One Cup", "EFL Cup", "capital-one-cup", "efl-cup", "eflcup", "EFL_CC_STACKED_REG_RGB")

  /**
   * The FA Cup.
   */
  case object FACP extends CupCompetition("FA Cup", "English FA Cup", "fa-cup", "facup")

  /**
   * The Championship.
   */
  case object FLC extends LeagueCompetition("Championship", "FLC", "FLD1")

  /**
   * The Championship play-offs.
   */
  case object FLCPO extends CupCompetition("Play-Offs", "FLD1 P/O", "FLC P/O")

  /**
   * Friendly matches.
   */
  case object FRIENDLY extends CupCompetition("Friendly")

  /**
   * The EUROPA League.
   */
  case object EUROPA extends CupCompetition("UEFA Europa League", "UEFA Europa League", "UEFA Europa League Qualifying", "europaleague")

  def apply(token: String): Competition = {
    values.find { competition =>
      competition.tokens.exists(competitionToken => token.contains(competitionToken))
    }.getOrElse(FRIENDLY)
  }

}
